---
title: "Exploration data NIOZ164, Discs EUX"
author: Maaike Goudriaan, MMB, NIOZ
output: html_document
date: "2023-08-09"
---
# Purpose
NIOZ164 is an illumina sequencing lane consisting of different samplesets. 
Here we focus on analysis of 16S amplicon data, from amplified DNA extracted from discs covered in polymer films that were incubated at 2 different locations in 2 different waterdepths (seafloor and watercolumn) in coastal Caribbean waters close to the island of St. EUstatius. THe tested polymers were PE, PP, PS, PEt and Nylon, PE-13C and PP-13C. Of each polymere there was an UV pretreated and a non-treated version. 

#Set-up markdown file and analysis environment
```{r setup, include = TRUE}
knitr::opts_chunk$set(echo = TRUE)

```

## What dir are we in and which files are there
```{r check-wd, r files in prject folder, results='hide'}
getwd()
print(dir()) 

``` 

## Load relevant libraries
```{r libaries, warning=FALSE,message=FALSE,error=FALSE, results='hide',fig.keep='all'}
library(devtools)
library(phyloseq)
library(grid)
library(tidyverse)
library(vegan)
library(rmarkdown)
library(knitr)
library(microbiome)
library(ggplot2)
library(ggpubr)
library(ggh4x)
library(ggrepel)

```

## Load data
```{r load data, results='hold', message=FALSE}
tax <- as.matrix(read.delim('../Data/representative_seq_set_tax_assignments.txt', row.names = 1, na.strings = c(" ")))
tax <- tax_table(tax)
otu <- read.delim('../Data/asvTable.txt', row.names = 1)
otu <- otu %>%  dplyr::select(-taxonomy) %>% as.matrix() 
otu <- otu_table(otu, taxa_are_rows = T)
map <- read.delim('../Data/Metadata_EUX_Discs.txt', row.names = 1, na.strings = c(""))

# Add extra combi columns to the map metadata matirx
map$Location_Habitat <- str_c(map$Location, "_", map$Habitat)
map$Polymer_Treatment <- str_c(map$Polymer, "_", map$Treatment)

map <- sample_data(map) 

```
# Create and check phyloseq
Total nr of reads in phyloseq 11,715,865
Lowest taxa sum is 0
Highest taxa sum is 402463
Mean taxa sum is 148.7767
```{r Create phyloseq object and check, results='hold', message=FALSE}
physeq_object = phyloseq::merge_phyloseq(otu, tax, map)
summarize_phyloseq(physeq_object) 


#store physeq object for later
saveRDS(physeq_object, "../Analysis/NIOZ164_physeq_object.rds")
```

```{r basic info physeq raw, echo=FALSE,warning=FALSE,message=FALSE,error=FALSE, results='hide',fig.keep='all'}
source("basic_info_physeq_object.R")
basic_info_physeq_object(physeq_object)

```

## Visualizing the read abundance per sample plot
```{r abundance plot, results='hide', message=FALSE, fig.keep='all'}
sample_sums <- sample_sums(physeq_object)
sample_sums_df <- data.frame(sum = sample_sums) %>% rownames_to_column("Sample_ID") %>% arrange(Sample_ID)
Sample_description <- data.frame(map) %>% rownames_to_column("Sample_ID") %>% arrange(Sample_ID)
sample_sums_df$Description <- Sample_description$Description
ggplot(sample_sums_df, aes(x = Description, y = sum)) +
  geom_bar(stat = "identity", fill = "blue") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(x = "Sample", y = "Sum of Counts", title = "Sample Sums of Phyloseq Object")

``` 
There is clearly a difference in readnumbers per sample. CBW19_B has low read abundance. 
2NTC are low, 1 is a bit high. 

## How much of each rank do we have, and how much is NA?
First we subset samples the remove the waterphase filters, which will not be included in the analyses
```{r calculations, results='hide', message=FALSE}
physeq_object.1 <- readRDS("../Analysis/NIOZ164_physeq_object.rds")
physeq_object <- subset_samples(physeq_object.1 , Phase %in% c('Disc', 'Wild'))
summarize_phyloseq(physeq_object) 
basic_info_physeq_object(physeq_object)

subset_samples(physeq_object, Location_Habitat == "Crooks_Castle_Pelagic") %>% summarize_phyloseq()
subset_samples(physeq_object, Location_Habitat == "Crooks_Castle_Benthic") %>% summarize_phyloseq()
subset_samples(physeq_object, Location_Habitat == "Charles_Brown_Pelagic") %>% summarize_phyloseq()
subset_samples(physeq_object, Location_Habitat == "Charles_Brown_Benthic") %>% summarize_phyloseq()


get_taxa_unique(physeq_object, "Kingdom")
subset_taxa(physeq_object, Kingdom == "Eukaryota") %>% summarize_phyloseq()
subset_taxa(physeq_object, Kingdom == "Archaea") %>% summarize_phyloseq()
subset_taxa(physeq_object, Kingdom == "NA") %>% summarize_phyloseq()

get_taxa_unique(physeq_object, "Phylum") %>% sort()
subset_taxa(physeq_object, Phylum == "NA") %>% summarize_phyloseq()
get_taxa_unique(physeq_object, "Order") %>% sort()
subset_taxa(physeq_object, Order == "NA") %>% summarize_phyloseq()
get_taxa_unique(physeq_object, "Family") %>% sort()
subset_taxa(physeq_object, Family == "NA") %>% summarize_phyloseq()
get_taxa_unique(physeq_object, "Family") %>% sort()
subset_taxa(physeq_object, Genus == "NA") %>% summarize_phyloseq()

```
## Readnumbers:
Total taxa: 78,748
Total reads: 9,880,102
Total samples: 68


### Per Habitat/Location
Crooks_Castle_Pelagic: 3,466,833 
Crooks_Castle_Benthic: 2,071,959
CB_Pelagic: 1,221,229
CB_Benthic: 2,144,505


### Kingdom level: 
- Total Eukaryotic reads = 689
- Total Archaeal reads = 69,960
- Total NA reads = 346

### Other levels
- Phylum level NA reads = 40,321
- Order level NA = 747,965
- Family level NA = 1,742,689
- Genus level NA = 4,347,209

## Turn absolute abundance data to a tidy tibble for rarefaction
```{r store make tibble,  results='hide', message=FALSE}
source("tidy_tibble_maker.R")
physeq_object <- subset_samples(physeq_object, Location != "NC")
tidy_physeq_abs <- tidy_tibble_maker(physeq_object)
write.csv(tidy_physeq_abs, '../Processed-Data/NIOZ165_EUX_discs_abs_data_unfiltered.csv')

```

# Rarefaction 
```{r Rarefaction curves, message=FALSE, results='hold'}
colnames(tidy_physeq_abs)
rownames(tidy_physeq_abs)
ASV.abs <-  tidy_physeq_abs%>%  select(Sample, Description, OTU, Abundance, Location, Habitat, Polymer, Treatment, Location_Habitat, Polymer_Treatment) %>%
  distinct() 
head(ASV.abs)
#Adjust the number of significant digets to get from vegan in rarefactions
old <- options(pillar.sigfig = 8)
options(old)

# Transform count table to get samples per row and ASV per column
# Not per se necessary, but this means we can use standard MARGIN in transfromation and ordination
ASV.ABS_df1<- ASV.abs %>% select(Sample, OTU, Abundance)  %>% mutate(across(c(OTU),factor)) %>% 
  pivot_wider(names_from = OTU, values_from = Abundance, values_fill = 0) %>% replace(is.na(.), 0)  %>% 
  arrange(Sample) %>% column_to_rownames(var = "Sample") %>% as.data.frame()
str(ASV.ABS_df1)
rownames(ASV.ABS_df1)
colnames(ASV.ABS_df1)

# Select the metadata we need for plotting
meta_df1 <- ASV.abs %>% select(Sample, Description, OTU, Abundance, Location, Habitat, Polymer, Treatment)  %>%
  mutate(across(c(OTU),factor)) %>% 
  pivot_wider(names_from = OTU, values_from = Abundance, values_fill = 0) %>% replace(is.na(.), 0)  %>% 
  arrange(Sample) %>% column_to_rownames(var = "Sample") %>% as.data.frame() %>% select(Location, Habitat, Polymer, Treatment)
rownames(meta_df1)
colnames(meta_df1)

# Find the lowest amount of sequences in all samples with this
# Can be used for rarefaction depth
min_seqs_1 <- ASV.abs%>%
  group_by(Sample) %>%
  summarize(n_seqs = sum(Abundance))  %>%
  summarize(min = min(n_seqs)) %>%
  pull(min)

#Remove the first column frpm the df for rarefaction 
ASV.ABS.df1 <- ASV.ABS_df1[,-1]

# Create the rarefaction curves for individual samples
# steps of 250 is high enough resolution for now
samples.rarecurve <- vegan::rarecurve(ASV.ABS.df1, step = 250) 
str(samples.rarecurve)

# Since it is a list of lists, with this piece of code we can turn it into a df for plotting
samp.rare.data <- map_dfr(samples.rarecurve[-99], bind_rows) %>% 
  bind_cols(meta_df1[-c(99),], Sample = rownames(ASV.ABS_df1[-c(99),]))  %>% 
  pivot_longer(-c("Sample", "Location", "Habitat", "Polymer", "Treatment")) %>% 
  drop_na() %>%
  mutate(n_seqs = as.numeric(str_replace(name, "N", ""))) %>% 
  select(-name)

head(samp.rare.data)
colnames(samp.rare.data)

# We add an extra column with samplenames to be used for the label
# Group by Sample and make sure there is only a label for the last value of the sample
samp.rare.data <- samp.rare.data %>% group_by(Sample) %>% mutate(label = if_else(value == max(value), as.character(Polymer), NA_character_))

rareplot.1 <- ggplot(samp.rare.data, aes(x = n_seqs, y = value, group = Sample)) +
  geom_line(aes(color = Treatment), linewidth = 1) +
  xlim(0, 250000) +
  # geom_vline(xintercept = 10916, color="black", linewidth = 1) + 
  facet_grid(~ Location + Habitat) +
  geom_label_repel(aes(label = label, color = Treatment), nudge_x = 1, nudge_y = -0.5, na.rm = T) +
  # scale_colour_manual() +
  xlab("# of sequences") + ylab("# of ASVs") +
  theme_classic()

rareplot.1

```

# Turn data to RA
```{r turn to RA,  results='hide', message=FALSE}
source("tidy_tibble_maker.R")
physeq_prefilt <- transform_sample_counts(physeq_object, function(x) x / sum(x) )

```

## Store the RA data
```{r store unfiltered data,  results='hide', message=FALSE}
#Save physeq for later
saveRDS(physeq_prefilt, "../Analysis/NIOZ164_physeq_object_RA_data.rds")
tidy_physeq_prefilt <- tidy_tibble_maker(physeq_prefilt)
write.csv(tidy_physeq_prefilt, '../Processed-Data/NIOZ165_EUX_discs_RA_data_unfiltered.csv')
```

# Check and correct taxonomy
Wonky_tonky_taxonomy removes ASVs unassigned on Kingdom and Phylum level, and removes Chloroplast and mitochondria.
After, we apply a loop on the taxonomy table to redefine "weird" taxonomy into a single common string unassigned, easier for plotting
```{r correct and filter,results='hold', message=FALSE}
source("wonky_tonky_taxonomy.R")
wonky_tonky_taxonomy(physeq_object)

# extract tax table as dataframe
taxo <- as.data.frame(physeq_object@tax_table)

# Loooop
for (i in 1:nrow(taxo)) {
  for (y in 1:ncol(taxo)) {
    if 
    (any(str_detect(taxo[i,y], c("uncultured","Uncultured","metagenome","Metagenome","unknown","Unknown","NA")))) {taxo[i,y] <- "unassigned" }
  }
} 

# re-define as tax table object
taxo <- tax_table(as.matrix(taxo))
# merge updated taxonomy
physeq_tax_correct <- merge_phyloseq(physeq_object@otu_table, taxo, map) 

# check the physeq object
summarize_phyloseq(physeq_tax_correct)

```

```{r basic info physeq post-filtering, echo=FALSE,warning=FALSE,message=FALSE,error=FALSE, results='hide',fig.keep='all'}
source("basic_info_physeq_object.R")
basic_info_physeq_object(physeq_object)

```

##Turn correctec data to RA
```{r}
physeq_tax_correct_RA <- transform_sample_counts(physeq_tax_correct, function(x) x / sum(x) )

#Save physeq for later
saveRDS(physeq_tax_correct_RA, "../Analysis/NIOZ164_physeq_object_RA_data_Tax_correct.rds")
```

## Check amounts of unassigned/removed taxa
```{r check, results='hold', message=FALSE}
Phyl <- subset_taxa(physeq_tax_correct, Phylum == "unassigned" )
#unassigned Phyla
summarize_phyloseq(Phyl)
mean(taxa_sums(Phyl))
sd(taxa_sums(Phyl))

Order <- subset_taxa(physeq_tax_correct, Order == "unassigned" )
#unassigned Orders
summarize_phyloseq(Order)
mean(taxa_sums(Order))
sd(taxa_sums(Order))

Family <- subset_taxa(physeq_tax_correct, Family == "unassigned" )
#unassigned Family
summarize_phyloseq(Family)
mean(taxa_sums(Family))
sd(taxa_sums(Family))

Genus <- subset_taxa(physeq_tax_correct, Genus == "unassigned" )
#unassigned Genus
summarize_phyloseq(Genus)
mean(taxa_sums(Genus))
sd(taxa_sums(Genus))

```

# Barplots on Kingdom, Phyla, Order level
What do we find in negative controls, do we need to filter/clean this?
```{r barplots, results='hold', message=FALSE}


physeq_tax_correct_RA <- readRDS("../Analysis/NIOZ164_physeq_object_RA_data_Tax_correct.rds")
physeq_plot <- subset_samples(physeq_tax_correct_RA , Phase %in% c('Disc', 'Wild'))
physeq_plot_p <- prune_samples(sample_sums(physeq_plot) > 0.05, physeq_plot)
summarize_phyloseq(physeq_plot_p)

bar_phylum <- plot_bar(physeq_plot_p, "Phylum", fill = "Phylum", facet_grid =  ~ Location)
bar_phylum

bar_order <- plot_bar(physeq_plot_p, "Order", fill = "Order", facet_grid =  ~ Location)
bar_order
```

#Prevalence plots
```{r prevalence plot}
microbiome::plot_taxa_prevalence(physeq_tax_correct, "Phylum") + 
  theme(legend.position = "none") +  
  labs(title = "Prevalence Prokaryotic Phyla")

#subset physeq
physeq_CC <- subset_samples(physeq_tax_correct, Location == "Crooks_Castle")
physeq_CB <- subset_samples(physeq_tax_correct, Location == "Charles_Brown")

physeq_Pel <- subset_samples(physeq_tax_correct, Habitat == "Pelagic")
physeq_Ben <- subset_samples(physeq_tax_correct, Habitat == "Benthic")

microbiome::plot_taxa_prevalence(physeq_CB, "Phylum") + 
  theme(legend.position = "none") +  
  labs(title = "Prevalence Prokaryotic Phyla Charles Brown")

microbiome::plot_taxa_prevalence(physeq_CC, "Phylum") + 
  theme(legend.position = "none") +  
  labs(title = "Prevalence Prokaryotic Phyla Crooks Castle")

```