---
title: "Curation and exploration data NIOZ164, Discs EUX"
author: Maaike Goudriaan, MMB, NIOZ
output: html_document
date: "2023-08-09"
---
# Purpose
NIOZ164 was an illumina sequencing lane consisting of different samplesets. 
Here we focus on analysis of 16S amplicon data, from amplified DNA extracted from discs covered in polymer films that were incubated at 2 different locations in 2 different waterdepths (seafloor and watercolumn) in coastal Caribbean waters close to the island of St. EUstatius. THe tested polymers were PE, PP, PS, PEt and Nylon, PE-13C and PP-13C. Of each polymere there was an UV pretreated and a non-treated version. 

#Set-up markdown file and analysis environment
```{r setup, include = TRUE}
knitr::opts_chunk$set(echo = TRUE)
```

## What dir are we in and which files are there
```{r check-wd, r files in prject folder, results='hide'}
getwd()
print(dir()) 
``` 

## Load relevant libraries
```{r libaries, results='hold', message=FALSE}
library(devtools)
library(phyloseq)
library(grid)
library(tidyverse)
library(vegan)
library(rmarkdown)
library(knitr)
library(microbiome)
```

## Load data
```{r load data, results='hold', message=FALSE}
tax <- as.matrix(read.delim('../Data/representative_seq_set_tax_assignments.txt', row.names = 1, na.strings = c(" ")))
tax <- tax_table(tax)
otu <- read.delim('../Data/asvTable.txt', row.names = 1)
otu <- otu %>%  select(-taxonomy) %>% as.matrix()
otu <- otu_table(otu, taxa_are_rows = T)
map <- read.delim('../Data/Metadata_EUX_Discs.txt', row.names = 1, na.strings = c(""))

# Add extra combi columns to the map metadata matirx
map$Location_Habitat <- str_c(map$Location, "_", map$Habitat)
map$Polymer_Treatment <- str_c(map$Polymer, "_", map$Treatment)
map$InputFileName <-NULL

map1 <- map %>%  mutate(Polymer_Isotope = if_else(Polymer %in% c("PE", "PP"), paste(Polymer, Isotope, sep = "-"), Polymer))
head(map1)

map1 <- sample_data(map1) 

#Directly load physeq_object
physeq_object <- readRDS("../Analysis/NIOZ164_physeq_object.rds")

```

## Check the data before filtering and pruning
"Total taxa is 78748"
"Total samples is 89"
"Lowest readnumber is 9080"
"Highest readnumber is 556374"
"Lowest taxa sum is 0"
"Highest taxa sum is 402463"

```{r Create phyloseq object and check, results='hide', message=FALSE}
source("basic_info_physeq_object.R")
basic_info_physeq_object(physeq_object)
```
# Check and correct taxonomy
First, pruning to get rid of ASVs with less then 20 reads in total, 
After, check if this leaves samples with only 1 ASV and remove these
Wonky_tonky_taxonomy removes ASVs unassigned on Kingdom and Phylum level, and removes Chloroplast and mitochondria.
After, we apply a loop on the taxonomy table to redefine "weird" taxonomy into a single communstring unassigned
```{r correct and filter,results='hold', message=FALSE}
physeq_pruned<- prune_taxa(taxa_sums(physeq_object) > 1, physeq_object) 
# physeq_pruned <- prune_samples(sample_sums(physeq_pruned) > 1, physeq_object)
summarize_phyloseq(physeq_pruned)

source("wonky_tonky_taxonomy.R")
wonky_tonky_taxonomy(physeq_pruned)

# extract tax table as dataframe
taxo <- as.data.frame(physeq_pruned@tax_table)

# Loooop
for (i in 1:nrow(taxo)) {
  for (y in 1:ncol(taxo)) {
    if 
    (any(str_detect(taxo[i,y], c("uncultured","Uncultured","metagenome","Metagenome","unknown","Unknown","NA")))) {taxo[i,y] <- "unassigned" }
  }
} 

# re-define as tax table object
taxo <- tax_table(as.matrix(taxo))
# merge updated taxonomy
physeq_filtered <- merge_phyloseq(physeq_object@otu_table, taxo, map1) 

```

## check physeq object after filtering
"Lowest readnumber is 9057"
"Highest readnumber is 556357"
"Lowest taxa sum is 21"
"Highest taxa sum is 402463"

[1] "3] Total number of reads = 11543430"

[[4]]
[1] "4] Average number of reads = 129701.460674157"

[[5]]
[1] "5] Median number of reads = 92126"

[[6]]
[1] "7] Sparsity = 0.947990086876778"

[[7]]
[1] "6] Any OTU sum to 1 or less? NO"

[[8]]
[1] "8] Number of singletons = 0"

[[9]]
[1] "9] Percent of OTUs that are singletons \n        (i.e. exactly one read detected across all samples)0"

[[10]]
[1] "10] Number of sample variables are: 11"

[[11]]
 [1] "Description"       "Location"          "Habitat"           "Polymer"           "Isotope"           "Backbone"          "Treatment"         "Method"            "Phase"            
[10] "Location_Habitat"  "Polymer_Treatment"

```{r check, results='hide', message=FALSE}
# check the physeq object
source("basic_info_physeq_object.R")
basic_info_physeq_object(physeq_filtered)
summarize_phyloseq(physeq_filtered)

```
## Transform data and store as a tidy tibble
Data is transformed to RA for plotting of commuity, and hellinger and rclr transformed for statistical analysis
```{r, store filtered data,  results='hide', message=FALSE}
source("tidy_tibble_maker.R")
tidy_filtered <- tidy_tibble_maker(physeq_filtered)
write.csv(tidy_filtered, '../Processed-Data/NIOZ165_EUX_discs_tidy_data_filtered.csv')

physeq_filtered_ra <- transform_sample_counts(physeq_filtered,function(x) x / sum(x) )
tidy_RA_filtered <- tidy_tibble_maker(physeq_filtered_ra)
write.csv(tidy_RA_filtered, '../Processed-Data/NIOZ165_EUX_discs_RA_data_filtered.csv')

physeq_filtered_hell <- microbiome::transform(physeq_filtered, 'hellinger' )
tidy_hell_filtered <- tidy_tibble_maker(physeq_filtered_hell)
write.csv(tidy_hell_filtered, '../Processed-Data/NIOZ165_EUX_discs_sqrt_RA_data_filtered.csv')

physeq_filtered_rclr <- microbiome::transform(physeq_filtered, 'rclr' )
tidy_rclr_filtered <- tidy_tibble_maker(physeq_filtered_rclr)
write.csv(tidy_rclr_filtered, '../Processed-Data/NIOZ165_EUX_discs_rclr_data_filtered.csv')

```

## Calculate the relative abundance per Taxonomic level per
```{r calculate RA in tibble, results='hide'}
tidy_RA <- tidy_filtered  %>% group_by(Description) %>% mutate(Sample_rel_abund = Abundance / sum(Abundance)) %>% #relative abundance of each otu per sample
  mutate(Sample_st_dev = sd(Sample_rel_abund)) %>% 
  ungroup() %>% 
  
  
  #_Kingdom_section__#
  group_by(Description, Kingdom) %>% 
  mutate(Kingdom_rel_abund_Sample = sum(Sample_rel_abund)) %>%  #Kingdom relative abundance per sample 
  ungroup() %>% 
  
  #_Phylum_section__#
  group_by(Description, Phylum) %>% 
  mutate(Phylum_rel_abund_Sample = sum(Sample_rel_abund)) %>%
  ungroup() %>% 
 
  #_Class_section__#
  group_by(Description, Class) %>% 
  mutate(Class_rel_abund_Sample = sum(Sample_rel_abund)) %>%
  ungroup() %>% 
  
  #_Order_section__#
  group_by(Description, Order) %>% 
  mutate(Order_rel_abund_Sample = sum(Sample_rel_abund)) %>%
  ungroup() %>% 
 
  #_Family_section__#
  group_by(Description, Family) %>% 
  mutate(Family_rel_abund_Sample = sum(Sample_rel_abund)) %>%
  ungroup() %>% 
  
  #_Genus_section__#
  group_by(Description, Genus) %>% 
  mutate(Genus_rel_abund_Sample = sum(Sample_rel_abund)) %>%
  ungroup() 

head(tidy_RA)
write.csv(tidy_RA, '../Processed-Data/NIOZ164_EUX_discs_tidy_RA_data_filtered.csv')
 
```

