---
title: "Curation and exploration data NIOZ164, Discs EUX"
author: Maaike Goudriaan, MMB, NIOZ
output: html_document
date: "2023-08-09"
---
# Purpose
NIOZ164 was an illumina sequencing lane consisting of different samplesets. 
Here we focus on analysis of 16S amplicon data, from amplified DNA extracted from discs covered in polymer films that were incubated at 2 different locations in 2 different waterdepths (seafloor and watercolumn) in coastal Caribbean waters close to the island of St. EUstatius. THe tested polymers were PE, PP, PS, PEt and Nylon, PE-13C and PP-13C. Of each polymere there was an UV pretreated and a non-treated version. 

#Set-up markdown file and analysis environment
```{r setup, include = TRUE}
knitr::opts_chunk$set(echo = TRUE)
```

## What dir are we in and which files are there
```{r check-wd, r files in prject folder, results='hide'}
getwd()
print(dir()) 
``` 

## Load relevant libraries
```{r libaries, results='hold', message=FALSE}
library(devtools)
library(phyloseq)
library(grid)
library(tidyverse)
library(vegan)
library(rmarkdown)
library(knitr)
library(microbiome)
```

## Load data
```{r load data, results='hold', message=FALSE}
map <- read.delim('../Data/Metadata_EUX_Discs.txt', row.names = 1, na.strings = c(""))

# Add extra combi columns to the map metadata matirx
map$Location_Habitat <- str_c(map$Location, "_", map$Habitat)
map$Polymer_Treatment <- str_c(map$Polymer, "_", map$Treatment)
map$InputFileName <-NULL

map1 <- map %>%  mutate(Polymer_Isotope = if_else(Polymer %in% c("PE", "PP"), paste(Polymer, Isotope, sep = "-"), Polymer))
head(map1)

map1 <- sample_data(map1) 

#Directly load physeq_object
physeq_object <- readRDS("../Analysis/NIOZ164_physeq_object_decontamed_tax.corrected.rds")

```

## Check the data before filtering and pruning
"Total taxa is 71573"
"Total samples is 68"
"Lowest readnumber is 7301"
"Highest readnumber is 539188"
"Total reads is 8,819,396
"Lowest taxa sum is 0"
"Highest taxa sum is 225,095"
"Number of singletons is 48,156"

```{r Create phyloseq object and check, results='hide', message=FALSE}
source("basic_info_physeq_object.R")
basic_info_physeq_object(physeq_object)
summarize_phyloseq(physeq_object)
```
# Prune physeq object
First, subset physeq for only discs and wild
Then, pruning to get rid of ASVs with less then 10 reads in total, 
After, check if this leaves samples with only 1 ASV and remove these
```{r correct and filter,results='hide', message=FALSE}
# Subsetting
physeq.wild.inc <- subset_samples(physeq_object, Phase %in% c('Disc', 'Wild')) 
summarize_phyloseq(physeq.wild.inc)
basic_info_physeq_object(physeq.wild.inc)

#Prunning
physeq.wild.inc.prune <- prune_samples(sample_sums(physeq.wild.inc) > 10, physeq.wild.inc)
# Remove all single singletons
physeq.wild.inc.prune.1 <- prune_taxa(taxa_sums(physeq.wild.inc.prune) > 1, physeq.wild.inc.prune) 
summarize_phyloseq(physeq.wild.inc.prune.1)
basic_info_physeq_object(physeq.wild.inc.prune.1)

100 * (8819396 - 8818707)/9880102
9880102 - 8818707

# Store the physeq object
saveRDS(physeq.wild.inc.prune.1, "../Analysis/NIOZ164_physeq_object_subset_decontamed_tax.corrected_pruned.rds")
```
# Check object after subsetting
"Total taxa is 71573"
"Total samples is 68"
"Lowest readnumber is 7301"
"Highest readnumber is 539184"
"Total reads is 8,819,396
"Lowest taxa sum is 0"
"Highest taxa sum is 225,095"
"Number of singletons is 48,156"

## check physeq object after Pruning
1] Min. number of reads = 7301
2] Max. number of reads = 549477
3] Total number of reads = 8,818,707
4] Average number of reads = 129686.87
Total taxa is 23,417
Total samples is 68

Singletons = 

## Visualize the read abundance per sample with plot
```{r abundance plot, results='hide', message=FALSE, fig.keep='all'}
sample_sums <- sample_sums(physeq.wild.inc.prune.1)
sample_sums_df <- data.frame(sum = sample_sums) %>% rownames_to_column("Sample_ID") %>% arrange(Sample_ID)
Sample_description <- data.frame(physeq.wild.inc.prune.1@sam_data) %>% rownames_to_column("Sample_ID") %>% arrange(Sample_ID) %>%  filter(Location != "NC")
sample_sums_df$Description <- Sample_description$Description
ggplot(sample_sums_df, aes(x = Description, y = sum)) +
  geom_bar(stat = "identity", fill = "blue") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(x = "Sample", y = "Sum of Counts", title = "Sample Sums after decontamination and pruning")

```
## Check amounts of unassigned/removed taxa
```{r, unassigned/removed taxa, results='hold', message=FALSE}
get_taxa_unique(physeq.wild.inc.prune.1, "Kingdom")
subset_taxa(physeq.wild.inc.prune.1, Kingdom == "Bacteria") %>% summarize_phyloseq()
subset_taxa(physeq.wild.inc.prune.1, Kingdom == "Archaea") %>% summarize_phyloseq()

Phyl <- subset_taxa(physeq.wild.inc.prune.1, Phylum == "unassigned" )
#unassigned Phyla
summarize_phyloseq(Phyl)
mean(taxa_sums(Phyl))
sd(taxa_sums(Phyl))

Order <- subset_taxa(physeq.wild.inc.prune.1, Order == "unassigned" )
#unassigned Orders
summarize_phyloseq(Order)
mean(taxa_sums(Order))
sd(taxa_sums(Order))

Family <- subset_taxa(physeq.wild.inc.prune.1, Family == "unassigned" )
#unassigned Family
summarize_phyloseq(Family)
mean(taxa_sums(Family))
sd(taxa_sums(Family))

Genus <- subset_taxa(physeq.wild.inc.prune.1, Genus == "unassigned" )
#unassigned Genus
summarize_phyloseq(Genus)
mean(taxa_sums(Genus))
sd(taxa_sums(Genus))

pct_bact = 100* (9024194/9095183)

```
Kingdom level: 
 - Total Bacterial reads = 9,024,194
 - Total Archaeal reads = 69,955
 
Other levels
 - Phylum level NA reads = 60,176
 - Order level NA = 910,216
 - Family level NA = 1,960,851
 - Genus level NA = 4,780,155

## Transform data and store as a tidy tibble
Data is transformed to RA for plotting of commuity, and hellinger and rclr transformed for statistical analysis
```{r, convert physeq to tibble,  results='hide', message=FALSE}
source("tidy_tibble_maker.R")
tidy_filtered <- tidy_tibble_maker(physeq.wild.inc.prune.1)

# Remove underscores for plotting
tidy_filtered.1 <- tidy_filtered %>% mutate(Location = case_when(
  Location == 'Crooks_Castle' ~ "Crooks Castle",
  Location == 'Charles_Brown' ~ "Charles Brown",
)
)

tidy_filtered.2 <- tidy_filtered.1 %>% mutate(Treatment = case_when(
 Treatment == 'no_UV' ~ "no UV",
 Treatment == 'UV' ~ "UV",
 
)
)

```

## Calculate the relative abundance per Taxonomic level per
```{r calculate RA in tibble, results='hide'}
tidy_RA <- tidy_filtered.2  %>% group_by(Description) %>% mutate(Sample_rel_abund = Abundance / sum(Abundance)) %>% #relative abundance of each otu per sample
  mutate(Sample_st_dev = sd(Sample_rel_abund)) %>% 
  ungroup() %>% 
  
  
  #_Kingdom_section__#
  group_by(Description, Kingdom) %>% 
  mutate(Kingdom_rel_abund_Sample = sum(Sample_rel_abund)) %>%  #Kingdom relative abundance per sample 
  ungroup() %>% 
  
  #_Phylum_section__#
  group_by(Description, Phylum) %>% 
  mutate(Phylum_rel_abund_Sample = sum(Sample_rel_abund)) %>%
  ungroup() %>% 
 
  #_Class_section__#
  group_by(Description, Class) %>% 
  mutate(Class_rel_abund_Sample = sum(Sample_rel_abund)) %>%
  ungroup() %>% 
  
  #_Order_section__#
  group_by(Description, Order) %>% 
  mutate(Order_rel_abund_Sample = sum(Sample_rel_abund)) %>%
  ungroup() %>% 
 
  #_Family_section__#
  group_by(Description, Family) %>% 
  mutate(Family_rel_abund_Sample = sum(Sample_rel_abund)) %>%
  ungroup() %>% 
  
  #_Genus_section__#
  group_by(Description, Genus) %>% 
  mutate(Genus_rel_abund_Sample = sum(Sample_rel_abund)) %>%
  ungroup() 

head(tidy_RA)
write.csv(tidy_RA, '../Processed-Data/NIOZ164_EUX_discs_RA_tidy_data_decontamed_tax.correct_pruned.csv')
 
```

